tests/chainer_tests/dataset_tests/tabular_tests/test_slice.py

```diff
@@ -0,0 +1,113 @@
+        {'indices': [i in {1, 3} for i in range(10)], 'expected_len': 2},
+        {'indices': [True] * 11, 'index_exception': ValueError},
+        {'indices': slice(3, None, -2), 'expected_len': 2}
+    ],
+    [
+        {'keys': None, 'expected_keys': ('a', 'b', 'c')},
+        {'keys': (1,), 'expected_keys': ('b',)},
+        {'keys': (3,), 'key_exception': IndexError},
+        {'keys': ('c',), 'expected_keys': ('c',)},
+        {'keys': ('d',), 'key_exception': KeyError},
+        {'keys': (-1, 'a'), 'expected_keys': ('c', 'a')},
+    ],
+    testing.product({
+        'get_examples_indices': [
+            None, [1], [1, 0], slice(0, 2, 1), slice(1, None, -1)],
+        'get_examples_key_indices': [None, (1,), (1, 0)],
+    }),
+))
+class TestSlice(unittest.TestCase):
+
+    def setUp(self):
+        self.exception = getattr(self, 'index_exception', None) \
+            or getattr(self, 'key_exception', None)
+
+        if isinstance(self.indices, list):
+            self.indices = [
+                index if isinstance(index, bool) else self.integer(index)
+                for index in self.indices]
+
+    def test_slice(self):
+        def callback(indices, key_indices):
+            if isinstance(self.indices, list) \
+                    or isinstance(self.get_examples_indices, list):
+                self.assertIsInstance(indices, list)
+            elif isinstance(self.indices, slice) \
+                    or isinstance(self.get_examples_indices, slice):
+                self.assertIsInstance(indices, slice)
+            else:
+                self.assertIsNone(indices)
+
+            if self.keys is None and self.get_examples_key_indices is None:
+                self.assertIsNone(key_indices)
+            else:
+                self.assertIsInstance(key_indices, tuple)
+
+        dataset = DummyDataset(
+            mode=self.mode, return_array=self.return_array, callback=callback)
+
+        if self.exception is not None:
+            with self.assertRaises(self.exception):
+                if self.keys is None:
+                    dataset.slice[self.indices]
+                else:
+                    dataset.slice[self.indices, self.keys]
+            return
+
+        if self.keys is None:
+            view = dataset.slice[self.indices]
+            data = dataset.data[:, self.indices]
+        else:
+            view = dataset.slice[self.indices, self.keys]
+            key_indices = [
+                {'a': 0, 'b': 1, 'c': 2}.get(key, key) for key in self.keys]
+            data = dataset.data[key_indices][:, self.indices]
+
+        self.assertIsInstance(view, TabularDataset)
+        self.assertEqual(len(view), self.expected_len)
+        self.assertEqual(view.keys, self.expected_keys)
+        self.assertEqual(view.mode, self.mode)
+
+        if self.get_examples_indices is not None:
+            try:
+                data = data[:, self.get_examples_indices]
+            except IndexError:
```

Comment:
I feel this logic is a bit dangerous because it is not obvious which combination of parameters are actually tested (there might be a possibility that one `indices` value is actually not tested at all; we need to verify it by carefully checking the parameterization...). Can we avoid it by improving the parameterization part? Same comments on key indices.

Reply:
At first, I customized parameterized part to do so.  However, it was somewhat hard to read. In this test, I'd like to check `slice` provides the same interfaces of `numpy.ndarray`. So using `numpy` to filter valid combinations.

Reply:
Hmm, ok. I'm happy if we could put this skipping logic to the upper of any assertions (at first glance, I was confused what will be checked in this test because assertions and this skipping logic are interleaving), but it is not a mandatory requirement for me.